ta/pkcs11/src/token_capabilities.c

```diff
@@ -0,0 +1,180 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2017-2020, Linaro Limited
+ */
+
+#include <assert.h>
+#include <pkcs11_ta.h>
+#include <string.h>
+#include <util.h>
+#include <tee_api.h>
+#include <tee_internal_api_extensions.h>
+
+#include "pkcs11_helpers.h"
+#include "token_capabilities.h"
+
+#define ALLOWED_PKCS11_CKFM	\
+	(PKCS11_CKFM_ENCRYPT | PKCS11_CKFM_DECRYPT |		\
+	 PKCS11_CKFM_DERIVE | PKCS11_CKFM_DIGEST |		\
+	 PKCS11_CKFM_SIGN | PKCS11_CKFM_SIGN_RECOVER |		\
+	 PKCS11_CKFM_VERIFY | PKCS11_CKFM_VERIFY_RECOVER |	\
+	 PKCS11_CKFM_GENERATE |	PKCS11_CKFM_GENERATE_KEY_PAIR |	\
+	 PKCS11_CKFM_WRAP | PKCS11_CKFM_UNWRAP)
+
+/*
+ * Definition of supported processings for a PKCS#11 mechanisms
+ * @id: Mechanism ID
+ * @flags: Valid PKCS11_CKFM_* for a mechanism as per PKCS#11
+ * @one_shot: true of mechanism can be used for a one-short processing
+ * @string: Helper string of the mechanism ID for debug purpose
+ */
+struct pkcs11_mechachism_modes {
+	uint32_t id;
+	uint32_t flags;
+	bool one_shot;
+#if CFG_TEE_TA_LOG_LEVEL > 0
+	const char *string;
+#endif
+};
+
+#if CFG_TEE_TA_LOG_LEVEL > 0
+#define MECHANISM(_label, _flags, _single_part)	\
+	{					\
+		.id = _label,			\
+		.one_shot = (_single_part),	\
+		.flags = (_flags),		\
+		.string = #_label,		\
+	}
+#else
+#define MECHANISM(_label, _flags, _single_part)	\
+	{					\
+		.id = _label,			\
+		.one_shot = (_single_part),	\
+		.flags = (_flags),		\
+	}
+#endif
+
+#define SINGLE_PART_ONLY	true
+
+/* PKCS#11 specificies permitted operation for each mechanism  */
+static const struct pkcs11_mechachism_modes pkcs11_modes[] = {
+	MECHANISM(PKCS11_CKM_AES_ECB,
+		  PKCS11_CKFM_ENCRYPT | PKCS11_CKFM_DECRYPT |
+		  PKCS11_CKFM_DERIVE |
+		  PKCS11_CKFM_WRAP | PKCS11_CKFM_UNWRAP, 0),
```

Comment:
Is this `0` intentional or would we better off with some define? It looks like it's really a boolean that should be supplied here.

Reply:
`0` is shorter than `false`. It is handy when using the long mechanism list...
well...
I will define a `ANY_STAGE`/`false` macro.